### PR TITLE
[graph_trainer] Coalesce chunked-loss reduce-scatter

### PR DESCRIPTION
**Pull request by wanlinan**
## Summary
Reduce redundant ReduceScatter communication in GraphTrainer chunked loss by replacing RS(g0) + ... + RS(gN) with RS(g0 + ... + gN).
The optimization operates within each microbatch, preserving existing cross-microbatch gradient accumulation. It is opt-in because changing the floating-point reduction order may change numerical results.
## Changes
- Add coalesce_chunked_loss_rs, disabled by default and requiring numerics_changing_optim.
- Match exclusive FP32/FP64 SUM ReduceScatter accumulation chains using existing parameter-gradient metadata.
- Accumulate gradients incrementally in an owned buffer, then issue a single ReduceScatter without mutating graph inputs or existing optimizer buffers.
- Run the pass after gradient-marker removal and dead-code elimination, before canonicalization and collective bucketing.

### CODE REVIEW

File: torchtitan/experiments/graph_trainer/configs.py [hunk 1]
```diff
@@ -150,6 +150,11 @@ class GraphTrainerCompileConfig(CompileConfig):
     """Enable passes that improve performance but may change numerics
     compared to the uncompiled path (e.g. RMSNorm Inductor fusion)."""
 
+    coalesce_chunked_loss_rs: bool = False
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/configs.py:153]: Can we reuse either this inductor config, or run the same pass in the GT pipeline?  https://github.com/pytorch/pytorch/pull/181309

**Comment by SherlockNoMad** [comment]
This should be fixed at tracing time, e.g. https://github.com/pytorch/torchtitan/pull/4182

instead of adhoc fix as a graph pass. 

I haven't got time to implement Tianyu's suggestion in the PR, but that's the agreed proper fix.

**Review by SherlockNoMad — CHANGES_REQUESTED** [review]
should be fixed at modeling code.
